spec/unit/pops/serialization/packer_spec.rb

```diff
@@ -0,0 +1,151 @@
+require 'spec_helper'
+require 'puppet/pops/serialization'
+
+module Puppet::Pops::Serialization
+[JSON].each do |packer_module|
+describe "the Puppet::Pops::Serialization when using #{packer_module.name}" do
+  let(:io) { StringIO.new }
+  let(:reader_class) { packer_module::Reader }
+  let(:writer_class) { packer_module::Writer }
+
+  def write(*values)
+    io.reopen
+    serializer = writer_class.new(io)
+    values.each { |val| serializer.write(val) }
+    serializer.finish
+    io.rewind
+  end
+
+  def read(count = nil)
+    @deserializer = reader_class.new(io)
+    count.nil? ? @deserializer.read : Array.new(count) { @deserializer.read }
+  end
+
+  context 'can write and read a Scalar' do
+    it 'String' do
+      val = 'the value'
+      write(val)
+      val2 = read
+      expect(val2).to be_a(String)
+      expect(val2).to eql(val)
+    end
+
+    it 'Integer' do
+      val = 32
+      write(val)
+      val2 = read
+      expect(val2).to be_a(Integer)
+      expect(val2).to eql(val)
+    end
+
+    it 'Float' do
+      val = 32.45
+      write(val)
+      val2 = read
+      expect(val2).to be_a(Float)
+      expect(val2).to eql(val)
+    end
+
+    it 'true' do
+      val = true
+      write(val)
+      val2 = read
+      expect(val2).to be_a(TrueClass)
+      expect(val2).to eql(val)
+    end
+
+    it 'false' do
+      val = false
+      write(val)
+      val2 = read
+      expect(val2).to be_a(FalseClass)
+      expect(val2).to eql(val)
+    end
+
+    it 'nil' do
+      val = nil
+      write(val)
+      val2 = read
+      expect(val2).to be_a(NilClass)
+      expect(val2).to eql(val)
+    end
+
+    it 'Regexp' do
+      val = /match me/
+      write(val)
+      val2 = read
+      expect(val2).to be_a(Regexp)
+      expect(val2).to eql(val)
+    end
+
+    it 'Time created by TimeFactory' do
+      # It does succeed on rare occasions, so we need to repeat
+      val = TimeFactory.now
+      write(val)
+      val2 = read
+      expect(val2).to be_a(Time)
+      expect(val2).to eql(val)
+    end
+
+    it 'Version' do
+      # It does succeed on rare occasions, so we need to repeat
+      val = Semantic::Version.parse('1.2.3-alpha2')
+      write(val)
+      val2 = read
+      expect(val2).to be_a(Semantic::Version)
+      expect(val2).to eql(val)
+    end
+
+    it 'VersionRange' do
+      # It does succeed on rare occasions, so we need to repeat
+      val = Semantic::VersionRange.parse('>=1.2.3-alpha2 <1.2.4')
+      write(val)
+      val2 = read
+      expect(val2).to be_a(Semantic::VersionRange)
+      expect(val2).to eql(val)
+    end
+
+    it 'can write and read Time created by TimeFactory' do
+      failed = false
+      # It does succeed on rare occasions, so we need to repeat
```

Comment:
You must mean "it fails on rare occasions" - and that is probably not true either since the intent here is to test that the implementation is robust and avoids brittle time methods that IF those were used then there may be a failure when iterating. (Otherwise, if we have an implementation that may fail - then that is not very good). Think you need to elaborate on the comment.